fvtest/tril/tril/parser.cpp

```diff
@@ -0,0 +1,482 @@
+                head_arg = createNodeArg("", buildNodeValue(*(tokenIt.getIt())), NULL);
+            } else {
+                throw TokenFailure("Incorrect type for argument, current value is " + tokenIt.getIt()->getValue());
+            }
+            if (head_argList == NULL) {
+                head_argList = head_arg;
+            } else {
+                appendSiblingArg(head_argList, head_arg);
+            }
+            ++tokenIt;
+        }
+    } catch (TokenFailure &f) {
+        std::cerr << f.what() << std::endl;
+    }
+    return head_argList;
+}
+
+
+ASTNode* buildAST(TokenIter &tokenIt) {
+    ASTNodeArg* head_argList = NULL;
+    ASTNode * head_child = NULL;
+    ASTNode * head_children = NULL;
+    ASTNode * head_siblings = NULL;
+
+    while (tokenIt.getIt()->getType() == Token::LPAREN) {
+        if (tokenIt.peek() == tokenIt.getEnd() || tokenIt.peek()->getType() != Token::ID) {
+            throw TokenFailure("The next token of LPAREN should be IDENTIFIER, while current value is " +  tokenIt.getIt()->getValue());
+        }
+        ++tokenIt;
+        char * node_name = new char[tokenIt.getIt()->getValue().size() + 1];
+        strcpy(node_name, tokenIt.getIt()->getValue().c_str());
+        ++tokenIt;
+        if (tokenIt.getIt()->getType() == Token::ID || tokenIt.getIt()->getType() == Token::DOUBLE_ || 
+            tokenIt.getIt()->getType() == Token::INT || tokenIt.getIt()->getType() == Token::STRING_ ||
+            tokenIt.getIt()->getType() == Token::HEXINT || tokenIt.getIt()->getType() == Token::ZERO) {
+            try {
+                head_argList = parseArgList(tokenIt);
+            } catch (TokenFailure &f) {
+                std::cerr << f.what() << std::endl;
+            }
+        }
+        head_child = buildAST(tokenIt);
+        if (head_children == NULL) {
+            head_children = head_child;
+        } else {
+            appendSiblingNode(head_children, head_child);
+        }
+        if (tokenIt.getIt()->getType() != Token::RPAREN && tokenIt.getIt() != tokenIt.getEnd()) {
+            throw TokenFailure("Missing RPAREN, while current value is " + tokenIt.getIt()->getValue());
+        }
+        ++tokenIt;
+        if (head_siblings == NULL) {
+            head_siblings = createNode(node_name, head_argList, head_children, NULL);
+        } else {
+            appendSiblingNode(head_siblings, createNode(node_name, head_argList, head_children, NULL));
+        }
+        head_children = NULL;
+        head_child = NULL;
+    }
+    return head_siblings;
+}
+
+
+ASTNode* parseFile(FILE *in) {
+    std::vector <Token> scanToken;
+    try {
+        char * line = NULL;
+        ssize_t read;
+        size_t len = 0;
+        while ((read = getline(&line, &len, in)) != -1) {
+            std::vector<Token> tokenLine = scan(line);
+            scanToken.insert(scanToken.end(), tokenLine.begin(), tokenLine.end());
+        }
```

Comment:
I used `fgetc` instead, and this could work, however, it's not effective enough. I also tried `fgets` be this required a specific length of chars to init the line, like `char line [100];`
Below is my new code.
``` 
             int c;
             std::string str;
             while ((c = std::fgetc(in)) != EOF) {
                  str += c;
             }
```